### PR TITLE
Retrieve and build dependency

### LINKED ISSUE #386: Implement Dependency RFC 0004 for  bundler
Implement [Dependency Management RFC Phase 1](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md) for `bundler`. Check out the RFC for more details and background.

### 1. Determine dependency source strategy
<details>
<summary>Background</summary>

When possible, dependencies should be used directly from upstream, rather than undergoing any additional compilation or modifications performed by Paketo-maintained code. For each dependency, the corresponding buildpack maintainer group will decide if the dependency can be used directly from upstream, and must identify the location from which the dependency will be pulled from. Some of the Paketo Java buildpacks perform directory stripping during the buildpack build process itself. This could be a viable alternative to performing directory modifications during the dependency management process for maintainers to consider.
</details>

- [x] Determine whether the dependency can be used directly from its upstream, rather than undergoing additional compilation/modification. 
* If using directly from upstream, note where dependencies will come from
* If compiled or modified, note the decision for this
- [x] Document decision in a language-family level RFC. This decision can be combined in the RFC with the decision made for other buildpacks in the language family.

### 2. Version retrieval and metadata generation code
Please refer to the [retrieval RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#1-version-retrieval-make-retrieve) on version retrieval for details around the API and background.

- [ ] Create retrieval code for the dependency
* Will live in the buildpack under: `dependency/retrieval/`
* Per [the RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#transfercombine-version-retrieval-and-metadata-generation), this will be a combination of (1) discovering new dependency versions based on the `buildpack.toml`, and (2) generating metadata for each new version.
* **Note** : Per [the RFC caveat](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#caveat-compiled-dependencies) - if the dependency is to be compiled, the `SHA256` and `URI` field from the metadata should be omitted in this step.

### 3. Compilation code (if needed)
If the dependency will be compiled/modified, then refer to the [compilation RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#2-compilation) for API details and background. 

- [ ] Create compilation action
* Will live in the buildpack under: `dependency/actions/compile/`

### 4. Dependency testing (optional)
It's up to maintainer discretion if the dependency will be tested. It's recommended to test the dependency if it's been compiled. Check out the [testing RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#3-test-dependency-make-test) for details.

- [ ] Add tests for dependency
* Will live in the buildpack under: `dependency/test.`

### 5. Makefile Setup
When using the generalized workflows for dependency management down the line, version retrieval and dependency testing will be executed via a Makefile in order to provide the workflow a standardized way to run the code, regardless of what language it was written in. Check out the [RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#overarching-plan) for what this should look like.

- [ ] Add Makefile
* Will live in the buildpack at: `dependency/Makefile`

### 6. Leveraging the new code
This issue serves to set up all the main logic for the dependency management. The work to actually leverage this code and migrate off of the dep-server will be completed in a separate issue once workflows and infrastructure is set up.

### PR DESCRIPTION
**Pull request by fg-j**
## Summary

Resolves #386

Uses logic similar to https://github.com/cloudfoundry/binary-builder/blob/69a6515f1ac1bbf89af51fb3e012e9e0d5181af6/recipe/bundler.rb to pull a bundler gem from upstream, strip out unnecessary components, and repackage it as a tar archive.

## Investigation of `gem install` during buildpack build phase
A brief investigation into alternate ways of installing bundler reveals that it's possible to `gem install` bundler in online and offline cases. Already, in the "compilation" code for this dependency, we use `gem install` with the `--local` flag and point it to a pre-downloaded bundler gem. (See the `gem install` [command reference](https://guides.rubygems.org/command-reference/) and [this StackOverflow](https://stackoverflow.com/questions/220176/how-can-i-install-a-local-gem) for more information.)Theoretically, the buildpack could do this at build time with a local copy of the upstream bundler gem, rather than un-tarring the premade bundler archive into a layer. Since the RFC for this dependency expressly stated that hosting our own versions of bundler dependencies was the preferred approach, I'll leave the implementation as-is unless @paketo-buildpacks/ruby-maintainers want to see it changed. 

## Checklist

* [ ] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [ ] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: dependency/tests/entrypoint [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  then
+    echo "missing bundle executable"
+    ls -al "${TESTDIR}"
+    exit 1
+  fi
+
+  # if ! [[ ${artifact} =~ [0-9]+\.[0-9]+\.[0-9]+ ]]; then
+  #   echo "No semver version in artifact name"
+  #   exit 1
+  # fi
```
Comment (ryanmoran) [inline dependency/tests/entrypoint:65]: Is this intentionally commented out?

File: dependency/retrieval/internal/metadata.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import "fmt"
+
+const cpePattern string = `cpe:2.3:a:bundler:bundler:%s:*:*:*:*:ruby:*:*`
+const sourceURI string = `https://rubygems.org/downloads/bundler-%s.gem`
+const depID string = "bundler"
+
+type ReleaseMetadata struct {
+	CPE string `json:"cpe"`
+	// TODO: decide whether to have DeprecationDate in the struct at all
+	// DeprecationDate string `json:"deprecation_date, omitempty"`
```
[the remark this answers is not in the document]
Comment (fg-j) [reply dependency/retrieval/internal/metadata.go:12]: I dug again to try and find a deprecation date, still no clear info on this. So I'm going to rip this out.

File: dependency/actions/compile/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+```
+
+2. Make the output directory:
+```
+mkdir <output dir>
+```
+
+3. Run compilation and use a volume mount to access it:
+```
+docker run -v <output dir>:/home compilation --version <version> --output-dir /home --target <target>
```
[the remark this answers is not in the document]
Comment (fg-j) [reply dependency/actions/compile/README.md:15]: Not as far as I'm aware. That's just what I was running when I did it on my local. I can make it more generic if that's clearer.

File: dependency/actions/compile/ubuntu.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,13 @@
+FROM ubuntu:18.04
+
+RUN apt-get -y update
+RUN apt-get -y install build-essential curl rubygems
+
+ARG cnb_uid=0
+ARG cnb_gid=0
+
+USER ${cnb_uid}:${cnb_gid}
```
[the remark this answers is not in the document]
Comment (fg-j) [reply dependency/actions/compile/ubuntu.Dockerfile]: Nope, shouldn't be.

File: dependency/actions/compile/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanmoran) [reply dependency/actions/compile/README.md]: Yeah. `/home` is meaningful for a user in Linux environments and so it made me think there was something special required to make the compilation work.

File: dependency/actions/compile/entrypoint [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  if [[ -z "${target:-}" ]]; then
+    echo "--target is required"
+    exit 1
+  fi
+
+  echo "Downloading bundler gem"
+
+  local gem_archive="${PROGDIR}/bundler.gem"
+  curl "https://rubygems.org/downloads/bundler-${version}.gem" \
```
Comment (robdimsdale) [inline dependency/actions/compile/entrypoint:56]: In general I like to use `curl --fail` (or `--fail-with-body` for newer versions).

Otherwise the script will continue on when `curl` failed to download something. That almost always results in a more complex error to debug later.

File: go.mod [hunk 6]
```diff
@@ -82,6 +82,7 @@ require (
 	github.com/opencontainers/go-digest v1.0.0 // indirect
 	github.com/opencontainers/image-spec v1.1.0-rc2 // indirect
 	github.com/opencontainers/runc v1.1.4 // indirect
+	github.com/package-url/packageurl-go v0.1.0 // indirect
```
Comment (robdimsdale) [inline go.mod:85]: When I run `go mod tidy` this line disappears, along with the equivalent in `go.sum`.

This makes senes to me, as the use of the `packageurl-go` library is in the `dependency` package which has its own go.mod file - we haven't added this package to the buildpack itself.

Could you validate this by running `go mod tidy` on your local copy and seeing if you get the same result?

**Review by robdimsdale — CHANGES_REQUESTED** [review]
I have one question and a couple of thoughts inline.

I have a couple of small suggestions which wouldn't block the PR, but if you're going to make any changes I might suggest we add them in.

File: dependency/actions/compile/entrypoint [hunk 7]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env bash
+
+set -eu
+set -o pipefail
```
[the remark this answers is not in the document]
Comment (robdimsdale) [reply dependency/actions/compile/entrypoint:4]: If you make this change here, I'd suggest making it to every script in this PR.

**Comment by sophiewigmore** [comment]
Hi! Just ran the validator tool and it didn't find the tests. I think the directory needs to be named `test` rather than `tests. The naming nitpick is unfortunately important for the purposes of the workflow looking up test code

File: go.mod (hunk 6 above)
Comment (fg-j) [reply go.mod:85]: I get the same result. I'll resolve that in a subsequent commit.

File: dependency/actions/compile/entrypoint (hunk 7 above)
[the remark this answers is not in the document]
Comment (fg-j) [reply dependency/actions/compile/entrypoint:4]: Added to the dependency scripts. Didn't add to things in `scripts/` since those will be overwritten by github config updates.

**Comment by sophiewigmore** [comment]
I can run retrieval, compilation, and tests now!

It looks like in the metadata, the `source-checksum` doesn't include the `algorithm`, if you can add that.

**Comment by sophiewigmore** [comment]
Don't forget to also remove the new `update-dependencies-from-metadata` workflow from the .syncignore / add the old `update-dependencies` to the .syncignore/delete the workflow.

**Comment by robdimsdale** [comment]
Also @fg-j we just got an automation bump to the bundler version in buildpack.toml so now there are some (minor) merge conflicts :(

**Comment by fg-j** [comment]
I'll add the algorithm for sure. `strip-components` is new. It's because the directory structure of the tar archive ends up a little different coming out of the compilation step. You can see that in the test scripting, when the tar archive is extracted, it's done with `--strip-components` as well.  I'll deal with the merge conflicts also, @robdimsdale!

File: buildpack.toml [hunk 8]
```diff
@@ -27,7 +27,7 @@ api = "0.7"
     sha256 = "d59b0711914a7876a10032d925811b9cfe87a1e9f87c82a861c3d69bb6857e93"
     source = "https://rubygems.org/downloads/bundler-1.17.3.gem"
     source_sha256 = "bc4bf75b548b27451aa9f443b18c46a739dd22ad79f7a5f90b485376a67dc352"
-    stacks = ["io.buildpacks.stacks.bionic"]
```
Comment (sophiewigmore) [inline buildpack.toml:30]: I'm slightly confused why only one dependency is marked with Jammy support. Also I see we're adding the Jammy stack support in this PR. Will there be a separate PR to regenerate all the other dependencies (so they're marked with Jammy support) and to add testing?

**Comment by sophiewigmore** [comment]
Besides that last comment - I'm happy with this PR as well.

File: buildpack.toml (hunk 8 above)
Comment (sophiewigmore) [reply buildpack.toml:30]: We can just leave this as-is if you're planning to essentially nuke the buildpack.toml so that all of the versions are regenerated with Jammy support

Comment (robdimsdale) [reply buildpack.toml:30]: I'd vote for rebuilding all dependencies with the new logic and jammy support. So I'd be happy to merge this as-is with the understanding we would rebuild the dependencies before shipping.
